### ISSUE TITLE
schemaOptions for embedded schemas

### ISSUE THREAD

**Issue by vemuez** [szokodiakos/typegoose#227]
What I do normally in mongoose is:

```
import { Schema, model } from 'mongoose';

const SubCategorySchema = new Schema({
    value: {
        type: String
    }
})

const CategorySchema = new Schema({
    value: {
        type: String,
        required: true
    },
    subCategories: [SubCategorySchema]
});

SubCategorySchema.set('toJSON', {
    virtuals: true,
    versionKey: false,
    transform: (doc, ret, options) =>
    {
        delete ret._id;
        return ret;
    }
})

CategorySchema.set('toJSON', {
    virtuals: true,
    versionKey: false,
    transform: (doc, ret, options) =>
    {
        delete ret._id;
        return ret;
    }
});

export const Category = model('Category', CategorySchema);
```

When the data comes through express and into my web app. The application prints the `id` rather than the `_id` for both `CategorySchema` and `SubCategorySchema` which is what I want. However, I can't seem to replicate this on typegoose. I can only manage to do it for `Category` by doing this:

```
import { Typegoose, prop, arrayProp } from 'typegoose';

import { ICategory, ISubCategory } from './category.interface';

export class SubCategory implements ISubCategory
{
    readonly id: string;

    @prop({ required: true })
    public value: string;
}

export class Category extends Typegoose implements ICategory
{
    readonly id: string;

    @prop({ required: true })
    public value: string;

    @arrayProp({ items: SubCategory, _id: false })
    public subCategories?: SubCategory[];
}

export const CategoryContext = new Category().getModelForClass(Category, {
    schemaOptions: {
        toJSON: {
            virtuals: true,
            versionKey: false,
            transform: (doc, ret, options) => {
                delete ret._id;
                return ret;
            }
        }
    }
});
```
Is this feature not implemented or have i missed it out from the docs? What other alternatives are there for this?

Stackoverflow link: https://stackoverflow.com/questions/53696331/schemaoptions-for-embedded-models-typegoose

**Comment by aoi-umi** [comment]
> 
> 
> What I do normally in mongoose is:
> 
> ```
> import { Schema, model } from 'mongoose';
> 
> const SubCategorySchema = new Schema({
>     value: {
>         type: String
>     }
> })
> 
> const CategorySchema = new Schema({
>     value: {
>         type: String,
>         required: true
>     },
>     subCategories: [SubCategorySchema]
> });
> 
> SubCategorySchema.set('toJSON', {
>     virtuals: true,
>     versionKey: false,
>     transform: (doc, ret, options) =>
>     {
>         delete ret._id;
>         return ret;
>     }
> })
> 
> CategorySchema.set('toJSON', {
>     virtuals: true,
>     versionKey: false,
>     transform: (doc, ret, options) =>
>     {
>         delete ret._id;
>         return ret;
>     }
> });
> 
> export const Category = model('Category', CategorySchema);
> ```
> 
> When the data comes through express and into my web app. The application prints the `id` rather than the `_id` for both `CategorySchema` and `SubCategorySchema` which is what I want. However, I can't seem to replicate this on typegoose. I can only manage to do it for `Category` by doing this:
> 
> ```
> import { Typegoose, prop, arrayProp } from 'typegoose';
> 
> import { ICategory, ISubCategory } from './category.interface';
> 
> export class SubCategory implements ISubCategory
> {
>     readonly id: string;
> 
>     @prop({ required: true })
>     public value: string;
> }
> 
> export class Category extends Typegoose implements ICategory
> {
>     readonly id: string;
> 
>     @prop({ required: true })
>     public value: string;
> 
>     @arrayProp({ items: SubCategory, _id: false })
>     public subCategories?: SubCategory[];
> }
> 
> export const CategoryContext = new Category().getModelForClass(Category, {
>     schemaOptions: {
>         toJSON: {
>             virtuals: true,
>             versionKey: false,
>             transform: (doc, ret, options) => {
>                 delete ret._id;
>                 return ret;
>             }
>         }
>     }
> });
> ```
> 
> Is this feature not implemented or have i missed it out from the docs? What other alternatives are there for this?

you can try to use [this package](https://www.npmjs.com/package/mongoose-ts-ua)

**Comment by vemuez** [comment]
@aoi-umi I've reverted back to mongoose. I'll have a try at that package later on. Thanks :)

**Comment by penumbra1** [comment]
I had a similar case. I thought I could call `setModelForClass` on the child schema and pass schema options to it (according to [the docs](https://github.com/szokodiakos/typegoose#api-documentation), it would overwrite the existing model for the child schema). I wonder why this doesn't work.

**Comment by Drageaux** [comment]
`schemaOptions` work for the top level doc only. When I queried the whole thing with embedded, schemaOptions like `_id: false` does not apply to the returned children object. This happens EVEN IF I set schemaOptions for the embedded classes

@hasezoey do you mind looking into this? 

```js

@ObjectType()
export class Boardextends Typegoose {
  @Field(type => [Square])
  @arrayProp({ items: Square, default: DEFAULT_BOARD })
  squares: Square[];
}

export const BoardModel = new Board().getModelForClass(Board, {
  schemaOptions: {
    _id: false,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
});

@ObjectType()
export class Square extends Typegoose {
}
export const SquareModel = new Square().getModelForClass(Square, {
  schemaOptions: {
    _id: false,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
});
```

**Comment by hasezoey** [comment]
@Drageaux first some things before i can answer: what is(when from typegoose, the imports would be good to know) `@Field`, `@ObjectType`, and what from these two do you want to embed | what is child, what is parent?

**Comment by diephil** [comment]
Anyone found a solution on this ? I'm having the same issue (sub-document with `_id` that I want to transform to `id`) :/

**Comment by hasezoey** [comment]
@diephil would #331 help?

**Comment by diephil** [comment]
Hello @hasezoey , I finally found a workaround for my issue.
In a context of an REST API endpoint (with NestJS), I'm calling a `parentSchema.toJSON()` before sending back my nested entities to the client.
The `toJSON` method searches deeply and replace all `_id` properties with `id`. Thank you for your suggestion

**Comment by hasezoey** [comment]
> schemaOptions for embedded schemas

this should work in 6.0.0 (not tested yet), thanks to `@modelOptions`

**Comment by hasezoey** [comment]
this can be closed because it is included in v6.0.0 (to an extend)

@Ben305
